src/tests/ftest/control/dmg_pool_query_ranks.py

```diff
@@ -80,58 +87,77 @@ def test_pool_query_ranks_mgmt(self):
-
-            self.log.debug("Waiting for pool to be rebuild")
-            self.pool.wait_for_rebuild_to_start()
-            self.pool.wait_for_rebuild_to_end()
+        exclude_rank = all_ranks[0]
+        suspect_rank = all_ranks[1]
+        suspect_ranks = [suspect_rank]
+        self.log.info("Starting excluding rank:%d all_ranks=%s", exclude_rank, all_ranks)
+        self.pool.exclude([exclude_rank])
+        enabled_ranks.remove(exclude_rank)
+        disabled_ranks = sorted(disabled_ranks + [exclude_rank])
 
-        self.random.shuffle(all_ranks)
-        self.log.info("Starting reintegrating ranks: all_ranks=%s", all_ranks)
-        for rank in all_ranks:
-            self.log.debug("Reintegrating rank %d", rank)
-
-            cmd_succeed = False
-            for _ in range(3):
-                try:
-                    result = self.pool.reintegrate(rank)
-                    cmd_succeed = True
-                    break
-                except CommandFailure:
-                    self.log.debug("dmg command failed retry")
-                time.sleep(3)
-
-            self.assertTrue(cmd_succeed, "pool reintegrate failed: {}".format(result))
-            enabled_ranks = sorted(enabled_ranks + [rank])
-            disabled_ranks.remove(rank)
-
-            self.log.debug("Checking enabled ranks state information")
-            data = self.dmg.pool_query(self.pool.identifier, show_enabled=True)
-            self.assertListEqual(
-                data['response'].get('enabled_ranks'), enabled_ranks,
-                "Invalid enabled_ranks field: want={}, got={}".format(
-                    enabled_ranks, data['response'].get('enabled_ranks')))
-            self.assertListEqual(
-                data['response'].get('disabled_ranks'), disabled_ranks,
-                "Invalid disabled_ranks field: want={}, got={}".format(
-                    disabled_ranks, data['response'].get('disabled_ranks')))
-
-            self.log.debug("Waiting for pool to be rebuild")
-            self.pool.wait_for_rebuild_to_start()
-            self.pool.wait_for_rebuild_to_end()
+        self.log.debug("Checking enabled ranks state information")
+        data = self.dmg.pool_query(self.pool.identifier, show_enabled=True)
+        self.assertListEqual(
+            data['response'].get('enabled_ranks'), enabled_ranks,
+            "Invalid enabled_ranks field: want={}, got={}".format(
+                enabled_ranks, data['response'].get('enabled_ranks')))
+        self.assertListEqual(
+            data['response'].get('disabled_ranks'), disabled_ranks,
+            "Invalid disabled_ranks field: want={}, got={}".format(
+                disabled_ranks, data['response'].get('disabled_ranks')))
+
+        self.log.debug("Waiting for pool to be rebuild")
+        self.pool.wait_for_rebuild_to_start()
+
+        # kill second rank.
+        self.server_managers[0].stop_ranks([suspect_rank], self.d_log)
+        time.sleep(20)
```

Comment:
Isn't there a better way than using fixed sleeps?? seems like we might be introducing intermittent failures by using them

Reply:
I agree with Tom. It would be better if we can deterministically decide when the rank has stopped. It's possible to check the corresponding log file, but I would need to dig up the relevant functions. Or do we think 20 seconds is almost always enough, even for various configs?

Reply:
Added a new helper for that.